processing/src/main/java/io/druid/query/extraction/LookupReferencesManager.java

```diff
@@ -0,0 +1,170 @@
+import com.metamx.common.logger.Logger;
+import io.druid.guice.ManageLifecycle;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This class provide a basic {@link LookupExtractorFactory} references manager.
+ * It allows basic operations fetching, listing, adding and deleting of {@link LookupExtractor} objects
+ * It is be used by queries to fetch the lookup reference.
+ * It is used by Lookup configuration manager to add/remove or list lookups configuration via HTTP or other protocols.
+ */
+
+@ManageLifecycle
+public class LookupReferencesManager
+{
+  private static final Logger LOGGER = new Logger(LookupReferencesManager.class);
+  private final ConcurrentMap<String, LookupExtractorFactory> lookupMap = new ConcurrentHashMap();
+  private final Object lock = new Object();
+  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  @LifecycleStart
+  public void start()
+  {
+    synchronized (lock) {
+      if (!started.getAndSet(true)) {
+        LOGGER.info("Started lookup factory references manager");
+      }
+    }
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    synchronized (lock) {
+      if (started.getAndSet(false)) {
+        LOGGER.info("Stopped lookup factory references manager");
+        for (String lookupName : lookupMap.keySet()) {
+          remove(lookupName);
+        }
+      }
+    }
+  }
+
+  /**
+   * @param lookupName             namespace of the lookupExtractorFactory object
+   * @param lookupExtractorFactory {@link LookupExtractorFactory} implementation reference.
+   *
+   * @return true if the lookup is added otherwise false.
+   *
+   * @throws IllegalStateException If the manager is closed or if start of lookup returns false.
+   */
+  public boolean put(String lookupName, final LookupExtractorFactory lookupExtractorFactory)
+  {
+    synchronized (lock) {
+      assertStarted();
+      if (!lookupExtractorFactory.start()) {
+        throw new ISE("start method returned false for lookup [%s]", lookupName);
+      }
+      return (null == lookupMap.putIfAbsent(lookupName, lookupExtractorFactory));
+    }
+  }
+
+  /**
+   * @param lookups {@link Map<String, LookupExtractorFactory>} containing all the lookup as one batch.
+   *
+   * @throws IllegalStateException if the manager is closed or if {@link LookupExtractorFactory#start()} returns false
+   */
+  public void put(Map<String, LookupExtractorFactory> lookups)
+  {
+    synchronized (lock) {
+      assertStarted();
+      for (Map.Entry<String, LookupExtractorFactory> entry : lookups.entrySet()) {
+        final String lookupName = entry.getKey();
+        final LookupExtractorFactory lookupExtractorFactory = entry.getValue();
+        if (!lookupExtractorFactory.start()) {
+          throw new ISE("start method returned false for lookup [%s]", lookupName);
```

Comment:
Throwing exception here will leave lookups in a weird state where some of the ones submitted were started but others were not. Suggest adding that information to the javadoc.